spark-etl/build.sbt

```diff
@@ -4,17 +4,29 @@ name := "geotrellis-spark-etl"
 libraryDependencies ++= Seq(
   jsonSchemaValidator,
   sparkCore % "provided",
-  scalatest % "test")
+  scalatest % "test"
+)
 
 test in assembly := {}
 
+assemblyShadeRules in assembly := {
+  val shadePackage = "com.azavea.shaded.demo"
+  Seq(
+    ShadeRule.rename("com.google.common.**" -> s"$shadePackage.google.common.@1")
+      .inLibrary("com.azavea.geotrellis" %% "geotrellis-cassandra" % Version.geotrellis).inAll,
+    ShadeRule.rename("io.netty.**" -> s"$shadePackage.io.netty.@1")
+      .inLibrary("com.azavea.geotrellis" %% "geotrellis-hbase" % Version.geotrellis).inAll,
+    ShadeRule.rename("com.fasterxml.jackson.**" -> s"$shadePackage.com.fasterxml.jackson.@1")
+      .inLibrary(jsonSchemaValidator).inAll,
+    ShadeRule.rename("org.apache.avro.**" -> s"$shadePackage.org.apache.avro.@1")
+      .inLibrary("com.azavea.geotrellis" %% "geotrellis-spark" % Version.geotrellis).inAll
+  )
+}
+
 assemblyMergeStrategy in assembly := {
   case s if s.startsWith("META-INF/services") => MergeStrategy.concat
-  case "reference.conf" => MergeStrategy.concat
-  case "application.conf" => MergeStrategy.concat
-  case "META-INF/MANIFEST.MF" => MergeStrategy.discard
-  case "META-INF\\MANIFEST.MF" => MergeStrategy.discard
-  case "META-INF/ECLIPSEF.RSA" => MergeStrategy.discard
-  case "META-INF/ECLIPSEF.SF" => MergeStrategy.discard
+  case "reference.conf" | "application.conf"  => MergeStrategy.concat
+  case "META-INF/MANIFEST.MF" | "META-INF\\MANIFEST.MF" => MergeStrategy.discard
+  case "META-INF/ECLIPSEF.RSA" | "META-INF/ECLIPSEF.SF" => MergeStrategy.discard
   case _ => MergeStrategy.first
 }
```

Comment:
FWIW, we used to have an assembly strategy like this for our Zeppelin integration, but it needed revising as some config files were getting dropped. Here's what we have now. Can't remember all the details, but it's closer to what the defaults are in the most recent `sbt-assembly` settings:

```scala
    assemblyMergeStrategy in assembly := {
      case "logback.xml" ⇒ MergeStrategy.singleOrError
      case x if Assembly.isConfigFile(x) ⇒ MergeStrategy.concat
      case PathList(ps @ _*) if Assembly.isReadme(ps.last) || Assembly.isLicenseFile(ps.last) ⇒
        MergeStrategy.rename
      case PathList("META-INF", xs @ _*) ⇒
        (xs map { _.toLowerCase }) match {
          case ("manifest.mf" :: Nil) | ("index.list" :: Nil) | ("dependencies" :: Nil) ⇒
            MergeStrategy.discard
          case ps @ (x :: _) if ps.last.endsWith(".sf") || ps.last.endsWith(".dsa") ⇒
            MergeStrategy.discard
          case "plexus" :: _ ⇒
            MergeStrategy.discard
          case "services" :: _ ⇒
            MergeStrategy.filterDistinctLines
          case ("spring.schemas" :: Nil) | ("spring.handlers" :: Nil) ⇒
            MergeStrategy.filterDistinctLines
          case ("maven" :: rest ) if rest.lastOption.exists(_.startsWith("pom")) ⇒
            MergeStrategy.discard
          case _ ⇒ MergeStrategy.deduplicate
        }

      case _ ⇒ MergeStrategy.deduplicate
    }
```

The first time you try it you're likely to get a lot of `deduplicate` errors... but this is good, as in our case found additional bugs relating to conflicting versions or identifying the need for additional shading rules.

Reply:
PS: One of the primary things that would get trashed with the `MergeStrategy.first` default was `META-INF/services` files. The version above makes sure they get merged.

Reply:
@metasim can you cllarify a bit where do you merge `META-INF/services` in your assembly reules? What problems did you have with the simplier (and not perfect) version?

I believe you had smth more than just an sbt-assembly build sbt file, it makes sense; and it's definitely a good point to add this generic assembly strategy.

Reply:
The relevant line is 
```
    case "services" :: _ ⇒
            MergeStrategy.filterDistinctLines
```

In our case we had to live *months* without an "Executors" page in the Spark UI associated with the Zeppelin driver. Finally got fed up and hunted it down to a particular Servlet endpoint not getting registered because of some SPI (I forget which one) registration getting smashed by some other one. Only happened when we pulled our über jar into Zeppelin; all was fine in `spark-submit`, etc.

Reply:
After the hours and hours we lost, I now think `MergeStrategy.first` or `MergeStrategy.discard` as wildcard defaults are evil. Use `MergeStrategy.deduplicate` to really understand what's getting dropped, and if you're OK with it then add a rule, filter the jar, or shade it.

Reply:
Well, it depends, we use merge by default for services meta-inf; it didn't work for you?